### PR TITLE
Enable ACPI on AArch64

### LINKED ISSUE #2178: Migrate ARM64 support to ACPI
Currently the ARM64 support used Device Tree requiring the dependency on the libfdt library. With the removal of virtio-mmio and GICv2 support it should be possible to move to ACPI only. This might give some hotplug support for free but most importantly simplify our code base.

@MrXinWang  @michael2012z  Happy for this change to go ahead?

### PR DESCRIPTION
**Pull request by michael2012z**
Partially fixes #2178.

This PR enabled ACPI on AArch64 and integrated with EDK2.

What's included:
- Updating some AML and FDT code
- Introducing EDK2 in CI
- Adding a basic test case

The PR is not totally ready. There are some issues to fix and discuss. Please see my comments.

### CODE REVIEW

File: vmm/src/device_manager.rs [hunk 1]
```diff
@@ -3727,6 +3727,48 @@ impl Aml for DeviceManager {
...
+                  }
+              }
+
+              Return (Buffer (One)
+              {
+                   0x00                                             // .
+              })
+        }
+        */
+        let uuid = aml::Uuid::new("E5C937D0-3553-4D7A-9117-EA4D19C3434D".to_string());
```
Comment (michael2012z) [inline vmm/src/device_manager.rs:3756]: We tried to refactor the code like https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2241/commits/e12ab49dc792b173f7b996110bb245ff1331df38, but it doesn't build.
@rbradford, @sboeuf , could you suggest a way to make it better?

File: scripts/run_integration_tests_aarch64.sh [hunk 2]
```diff
@@ -8,9 +8,73 @@ export BUILD_TARGET=${BUILD_TARGET-aarch64-unknown-linux-gnu}
 
 WORKLOADS_DIR="$HOME/workloads"
 WORKLOADS_LOCK="$WORKLOADS_DIR/integration_test.lock"
+EDK2_BUILD_DIR="$WORKLOADS_DIR/edk2_build"
 
 mkdir -p "$WORKLOADS_DIR"
 
+build_edk2() {
+    EDK2_REPO="https://github.com/jongwu/edk2.git"
```
Comment (michael2012z) [inline scripts/run_integration_tests_aarch64.sh:16]: Now the EDK2 code is still on review and discussion, not ready on master.

@rbradford Could you help create a branch for this in CH fork of edk2?

File: vmm/Cargo.toml [hunk 3]
```diff
@@ -28,7 +28,7 @@ event_monitor = { path = "../event_monitor" }
 hypervisor = { path = "../hypervisor" }
 lazy_static = "1.4.0"
 libc = "0.2.95"
-linux-loader = { version = "0.3.0", features = ["elf", "bzimage", "pe"] }
+linux-loader = { git = "https://github.com/jongwu/linux-loader", branch = "uefi", version = "0.3.0", features = ["elf", "bzimage", "pe"] }
```
Comment (michael2012z) [inline vmm/Cargo.toml:31]: Currently we are using some source code that is on review. And the code there is a bit out-of-date. @jongwu will rebase it soon.

File: scripts/run_integration_tests_aarch64.sh (hunk 2 above)
Comment (rbradford) [reply scripts/run_integration_tests_aarch64.sh:16]: @michael2012z You should have access now to do that.

Comment (michael2012z) [reply scripts/run_integration_tests_aarch64.sh:16]: Hi, @rbradford, sorry I still can't.
I am in group `cloud-hypervisor-write`, but the permission setting for `cloud-hypervisor/edk2` there is `Read`. Could you change it `Write`? Thanks.

Comment (sboeuf) [reply scripts/run_integration_tests_aarch64.sh:16]: Can you please try it now @michael2012z ? I've update the group setting to `Write`.

Comment (michael2012z) [reply scripts/run_integration_tests_aarch64.sh:16]: Thank you, it works. :)

File: vmm/Cargo.toml [hunk 4]
```diff
@@ -28,7 +28,7 @@ event_monitor = { path = "../event_monitor" }
 hypervisor = { path = "../hypervisor" }
 lazy_static = "1.4.0"
 libc = "0.2.95"
-linux-loader = { version = "0.3.0", features = ["elf", "bzimage", "pe"] }
+linux-loader = { git = "https://github.com/michael2012z/linux-loader", branch = "aarch64-uefi", features = ["elf", "bzimage", "pe"] }
```
Comment (michael2012z) [inline vmm/Cargo.toml:31]: This is temporary for introducing the new interface in `vm.rs`.

File: tests/integration.rs [hunk 5]
```diff
@@ -355,7 +355,7 @@ mod tests {
 
             let mut fw_path = workload_path;
             #[cfg(target_arch = "aarch64")]
-            fw_path.push("Image");
+            fw_path.push("CLOUDHV_EFI.fd");
```
Comment (sboeuf) [inline tests/integration.rs:358]: Wait this is a big change, you've moving all `aarch64` tests from direct kernel boot to EFI firmware boot. Is this intentional?

File: vmm/src/vm.rs [hunk 6]
```diff
@@ -875,6 +877,21 @@ impl Vm {
             None,
         ) {
             Ok(entry_addr) => entry_addr,
+            // Try to load the binary as kernel PE file at first.
```
Comment (michael2012z) [inline vmm/src/vm.rs:880]: This is related to a `linux-loader` proposal https://github.com/rust-vmm/linux-loader/pull/75.

To load the format-less UEFI binary, we figured out this logic.
@rbradford @sboeuf could you please take a look and give some comments?

File: vmm/Cargo.toml (hunk 4 above)
Comment (sboeuf) [reply vmm/Cargo.toml:31]: Please mark this PR as a draft until your PR on the linux loader gets merged.

File: tests/integration.rs [hunk 7]
```diff
@@ -2038,6 +2038,41 @@ mod tests {
...
+        fn test_edk2_acpi_launch() {
+            let focal = UbuntuDiskConfig::new(FOCAL_IMAGE_NAME.to_string());
+
+            vec![Box::new(focal)].drain(..).for_each(|disk_config| {
+                let guest = Guest::new(disk_config);
+
+                let mut child = GuestCommand::new(&guest)
+                    .args(&["--cpus", "boot=1"])
+                    .args(&["--memory", "size=512M"])
+                    .args(&["--kernel", guest.fw_path.as_str()])
```
Comment (sboeuf) [inline tests/integration.rs:2052]: I'd suggest you provide `CLOUDHV_EFI.fd` here, so that you don't have to modify the behavior for all existing tests.

File: tests/integration.rs (hunk 5 above)
Comment (michael2012z) [reply tests/integration.rs:358]: I intended, but it doesn't impact much on aarch64. Because in most aarch64 test cases, kernel was set with `direct_kernel_boot_path()`.

File: tests/integration.rs (hunk 7 above)
Comment (michael2012z) [reply tests/integration.rs:2052]: I agree this. Now that this is a edk2 specific test case, it's reasonable to use the edk2 specific binary `CLOUDHV_EFI.fd` in this case only.

File: tests/integration.rs (hunk 5 above)
Comment (michael2012z) [reply tests/integration.rs]: It was changed back.

File: vmm/src/vm.rs (hunk 6 above)
Comment (michael2012z) [reply vmm/src/vm.rs:889]: Added the load function in `arch/`.

File: vmm/src/device_manager.rs [hunk 8]
```diff
@@ -3670,6 +3670,48 @@ impl Aml for DeviceManager {
...
+        let dsm_eq1 = aml::Equal::new(&aml::Arg(0), &uuid);
+        let dsm_rt11 = aml::Return::new(&buf1);
+        let dsm_if11 = aml::If::new(&dsm_eq11, vec![&dsm_rt11]);
+        let dsm_rt12 = aml::Return::new(&aml::ZERO);
+        let dsm_eq12 = aml::Equal::new(&aml::Arg(2), &0x05u8);
+        let dsm_if12 = aml::If::new(&dsm_eq12, vec![&dsm_rt12]);
+        let dsm_if1 = aml::If::new(&dsm_eq1, vec![&dsm_if11, &dsm_if12]);
+        let dsm_rt2 = aml::Return::new(&buf2);
+        let dsm = aml::Method::new("_DSM".into(), 4, false, vec![&dsm_if1, &dsm_rt2]);
+        pci_dsdt_inner_data.push(&dsm);
```
Comment (MrXinWang) [inline vmm/src/device_manager.rs:3713]: @rbradford @sboeuf I wonder if this part will affect the x86. Could you please share your opinion about this? (It seems that we need this for Arm but for x86 personally I am not sure)

**Review by MrXinWang — CHANGES_REQUESTED** [review]

File: vmm/src/device_manager.rs (hunk 1 above)
Comment (rbradford) [reply vmm/src/device_manager.rs]: @michael2012z I pushed a commit into your branch to simplify. Please review and squash if it's still correct.

File: acpi_tables/src/aml.rs [hunk 9]
```diff
@@ -262,6 +263,38 @@ impl Aml for EisaName {
...
+// UUID (Universally Unique IDentifiers)
+pub struct Uuid {
+    pub data: Vec<u8>,
+}
+
+// Converts an ASCII UUID or GUID string to an encoded 128-bit Buffer object.
+// The UUID string is in the format of aabbccdd-eeff-gghh-iijj-kkllmmnnoopp,
+// where aa – pp are one byte hexadecimal numbers, made up of hexadecimal digits.
+impl Uuid {
+    pub fn new(str: String) -> Self {
```
Comment (rbradford) [inline acpi_tables/src/aml.rs:275]: This isn't ACPI specific. This is just a standard UUID definition. So this doesn't belong in this crate. Use the `uuid` crate in vmm to parse the string.

File: vmm/src/device_manager.rs (hunk 1 above)
Comment (michael2012z) [reply vmm/src/device_manager.rs]: Fantastic! Thank you very much for improving this. I squashed into Jongwu's original commit.

File: vmm/src/device_manager.rs (hunk 8 above)
Comment (michael2012z) [reply vmm/src/device_manager.rs:3733]: The related definition is in ACPI spec v6.3 Ch 9.1.1 and PCI Firmware spec v3.3 Ch 4.6.1. Looks like it's common among architectures.

File: acpi_tables/src/aml.rs (hunk 9 above)
Comment (michael2012z) [reply acpi_tables/src/aml.rs:275]: Aha, yes, I removed it. The endian of the UUID is a little bit specific (mixed-endian) but still standard (https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding).  We can use `uuid` crate.

File: vmm/src/device_manager.rs (hunk 8 above)
Comment (MrXinWang) [reply vmm/src/device_manager.rs]: From the CI result, it seems the change does not break x86, so I am ok with that.

**Review by MrXinWang — APPROVED** [review]

File: vmm/Cargo.toml [hunk 10]
```diff
@@ -51,6 +51,7 @@ vm-memory = { version = "0.5.0", features = ["backend-mmap", "backend-atomic", "
 vm-migration = { path = "../vm-migration" }
 vm-virtio = { path = "../vm-virtio" }
 vmm-sys-util = { version = ">=0.5.0", features = ["with-serde"] }
+uuid = "0.8"
```
Comment (rbradford) [inline vmm/Cargo.toml:54]: nit: We try and keep dependences in this file in alphabetical order.

File: vmm/src/acpi.rs [hunk 11]
```diff
@@ -378,6 +378,8 @@ pub fn create_acpi_tables(
     guest_mem
         .write_slice(dsdt.as_slice(), dsdt_offset)
         .expect("Error writing DSDT table");
+    #[cfg(target_arch = "aarch64")]
+    tables.push(dsdt_offset.0);
```
Comment (rbradford) [inline vmm/src/acpi.rs:382]: Why is this aarch64 specific? Does the DSDT actually need to be in the XSDT?

**Review by rbradford — APPROVED** [review]

File: vmm/src/acpi.rs (hunk 11 above)
Comment (michael2012z) [reply vmm/src/acpi.rs:382]: Thanks for reminding this.
This is a early test code. We need a fix on EDK2 code.
I will change it back when EDK2 is ready. Before that I will change the PR draft, in case it's merged unexpectedly.

File: vmm/Cargo.toml (hunk 10 above)
Comment (michael2012z) [reply vmm/Cargo.toml]: Done.

File: vmm/src/acpi.rs (hunk 11 above)
Comment (michael2012z) [reply vmm/src/acpi.rs]: This is removed now. Thanks for @jongwu fixing the issue in EDK2.

The PR is ready for review again.

File: scripts/run_integration_tests_aarch64.sh [hunk 12]
```diff
@@ -308,6 +308,17 @@ echo "Integration test on FDT finished with result $RES."
...
     cargo build --all --release $features_build_acpi --target $BUILD_TARGET
+    strip target/$BUILD_TARGET/release/cloud-hypervisor
+    strip target/$BUILD_TARGET/release/vhost_user_net
+    strip target/$BUILD_TARGET/release/ch-remote
+
+    # Enable KSM with some reasonable parameters so that it won't take too long
+    # for the memory to be merged between two processes.
+    sudo bash -c "echo 1000000 > /sys/kernel/mm/ksm/pages_to_scan"
+    sudo bash -c "echo 10 > /sys/kernel/mm/ksm/sleep_millisecs"
+    sudo bash -c "echo 1 > /sys/kernel/mm/ksm/run"
```
Comment (sboeuf) [inline scripts/run_integration_tests_aarch64.sh:319]: Why did you add these? Are you using KSM?

**Review by sboeuf — APPROVED** [review]

File: scripts/run_integration_tests_aarch64.sh (hunk 12 above)
Comment (michael2012z) [reply scripts/run_integration_tests_aarch64.sh:319]: I simply copied these from lines above. I cannot remember why they were there. Should I remove them?

Comment (michael2012z) [reply scripts/run_integration_tests_aarch64.sh]: Ah, got your meaning. You mean they are not needed for that ACPI test case, right?

Comment (sboeuf) [reply scripts/run_integration_tests_aarch64.sh]: Yes that's what I meant :)

**Comment by sboeuf** [comment]
@michael2012z @MrXinWang anything else needed on this PR?
If we're good to go, please go ahead and merge it when the CI turns green :)

**Comment by MrXinWang** [comment]
> @michael2012z @MrXinWang anything else needed on this PR?
> If we're good to go, please go ahead and merge it when the CI turns green :)

Thanks @sboeuf. As both you and @rbradford approved this PR, I will merge it then ^^
